influxdb3_server/src/http/v1.rs

```diff
@@ -140,23 +209,81 @@ pub struct QueryError(#[from] anyhow::Error);
 struct QueryResponse {
     results: Vec<StatementResponse>,
     #[serde(skip_serializing)]
-    pretty: bool,
+    format: QueryFormat,
 }
 
 /// Convert [`QueryResponse`] to [`Bytes`] for `hyper`'s [`Body::wrap_stream`] method
 impl From<QueryResponse> for Bytes {
     fn from(s: QueryResponse) -> Self {
-        if s.pretty {
-            serde_json::to_vec_pretty(&s)
-        } else {
-            serde_json::to_vec(&s)
+        /// Convert a [`QueryResponse`] to a JSON byte vector.
+        ///
+        /// This function serializes the `QueryResponse` to JSON. If the format is
+        /// `Pretty`, it will produce human-readable JSON, otherwise it produces
+        /// compact JSON.
+        fn to_json(s: QueryResponse) -> Vec<u8> {
+            if s.format.is_pretty() {
+                serde_json::to_vec_pretty(&s)
+                    .expect("Failed to serialize QueryResponse to pretty JSON")
+            } else {
+                serde_json::to_vec(&s).expect("Failed to serialize QueryResponse to JSON")
+            }
+        }
+
+        /// Convert a [`QueryResponse`] to a CSV byte vector.
+        ///
+        /// This function serializes the `QueryResponse` to CSV format. It dynamically
+        /// extracts column names from the first series and writes the header and data
+        /// rows to the CSV writer.
+        fn to_csv(s: QueryResponse) -> Vec<u8> {
+            let mut wtr = csv::Writer::from_writer(vec![]);
+            // Extract column names dynamically from the first series
+            let mut headers = vec!["name".to_string(), "tags".to_string()];
```

Comment:
Is it possible to avoid `to_string()` here, and the `clone`ing below? The `csv` writer accepts writes with borrowed data.

I see below that the row values, which are `serde_json::Value`s are converted to strings, so maybe that is unavoidable below, but calls to `wtr.write_record` should be independent so perhaps it could be done here...

Reply:
Done